lib/internal/watchdog.js

```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+const { JsSigintWatchdog } = internalBinding('watchdog');
+
+class SigintWatchdog extends JsSigintWatchdog {
+  _started = false;
+  _effective = false;
+  _onNewListener = (eve) => {
+    if (eve === 'SIGINT' && this._effective) {
+      super.stop();
+      this._effective = false;
+    }
+  };
+  _onRemoveListener = (eve) => {
+    if (eve === 'SIGINT' && process.listenerCount('SIGINT') === 0 &&
+        !this._effective) {
+      super.start();
+      this._effective = true;
+    }
+  }
+
+  start() {
+    if (this._started) {
+      return;
+    }
+    this._started = true;
+    /**
+     * Prepend sigint newListener to remove stop watchdog before signal wrap
+     * been activated. Also make sigint removeListener been ran after signal
+     * wrap been stopped.
+     */
+    process.prependListener('newListener', this._onNewListener);
```

Comment:
The event handler needs to be bound to the instance, otherwise `this` in the handler will point to `process`.

Reply:
Alternatively, instead of using a class for all of this, just use local scope for all of the state since only one instance of the class is being created anyway. That way you don't have to worry about `this` values being correct or performance issues with classes.

Reply:
They are bound to the instance.

Reply:
@addaleax Can you explain? Currently `this` in `_onNewListener()` will point to `process` (instead of the `SigintWatchdog` instance) because event handlers are always called in the context of the emitter. That does not seem like the intended behavior here.

Reply:
`_onNewListener()` is an arrow function, which means that `this` is automatically bound to the `SigintWatchdog` instance.

If you’d prefer to avoid this syntax for readability reasons, I think that’s fair – it’s a bit unusual, in part because it’s a new language feature, even if it works fine here, and I definitely had to read this more than once to make sure it’s right.

(I wouldn’t give performance considerations too much weight here because this is not a hot path, it only runs once per process at most.)